File: sort_algorithm/radix_sort.go
package sort_algorithm

import (
	"fmt"
	"math"
)

type RadixSort []int

func (r *RadixSort) SortImpl() {
	defer fmt.Println(r)
	if len(*r) < 2 {
		return
	}

	max := (*r)[0]
	for _, rr := range *r {
		if max < rr {
			max = rr
		}
	}
	num := 0
	for max > 0 {
		num++
		max /= 10
	}

	const BucketNum = 10

	buckets := make([][]int, BucketNum)
	for i := 1; i <= num; i++ {
		for _, rr := range *r {
			radio := rr / int(math.Pow(10, float64(i)-1)) % 10
			buckets[radio] = append(buckets[radio], rr)
		}

		*r = []int{}
		for j := 0; j < 10; j++ {
			*r = append(*r, buckets[j]...)
			buckets[j] = []int{}
		}
	}
}

File: java/sort_algorithm/impl/RadixSort.java
package sort_algorithm.impl;
import sort_algorithm.Sort;
import java.util.*;

public class RadixSort implements Sort {
    @Override
    public void sort(int[] arr) {
        int max = Arrays.stream(arr).max().orElse(arr[0]);
        int len = (max+"").length();
        final int BucketNum = 10;
        List<Integer>[] buckets = new List[BucketNum];
        for (int i = 0; i < buckets.length; i++) {
            buckets[i] = new LinkedList<>();
        }
        for (int i = 1; i <= len; i++) {
            for (int k : arr) {
                int index = (int) (k / Math.pow(10, i - 1) % BucketNum);
                buckets[index].add(k);
            }
            int index = 0;
            for (List<Integer> bucket : buckets) {
                for (Integer integer : bucket) {
                    arr[index++] = integer;
                }
            }
            for (List<Integer> bucket : buckets) {
                bucket.clear();
            }
        }
        System.out.println(Arrays.toString(arr));
    }
}
